src/import-export/csv-imp/gnc-tx-import.cpp

```diff
@@ -0,0 +1,960 @@
+    bool have_date = (trans_props.find (GncTransPropType::DATE) != trans_props.end());
+    bool have_amount = ((trans_props.find (GncTransPropType::DEPOSIT) != trans_props.end()) ||
+                        (trans_props.find (GncTransPropType::WITHDRAWAL) != trans_props.end()) ||
+                        (trans_props.find (GncTransPropType::BALANCE) != trans_props.end()));
+
+    std::string error_message {""};
+    if (!have_date)
+        error_message += N_("No date column.");
+    if (!have_amount)
+    {
+        if (!have_date)
+            error_message += "\n";
+        error_message += N_("No balance, deposit, or withdrawal column.");
+    }
+    if (!have_date || !have_amount)
+        *error = g_strdup (error_message.c_str());
+
+    return have_amount && have_date;
+}
+
+/** Create a Transaction from a map of transaction properties.
+ * Note: this function assumes all properties in the map have been verified
+ *       to be valid. No further checks are performed here other than that
+ *       the required properties are in the map
+ * @param transprops The map of transaction properties
+ * @param error Contains an error on failure
+ * @return On success, a GncCsvTransLine; on failure, the trans pointer is NULL
+ */
+static GncCsvTransLine* trans_properties_to_trans (prop_map_t& trans_props, gchar** error)
+{
+
+    if (!trans_properties_verify_essentials(trans_props, error))
+        return NULL;
+
+    auto property = trans_props.find (GncTransPropType::ACCOUNT)->second;
+    auto account = dynamic_cast<GncTransPropImpl<Account*>*>(property.get())->value;
+
+    GncCsvTransLine* trans_line = g_new (GncCsvTransLine, 1);
+
+    /* The balance is 0 by default. */
+    trans_line->balance_set = false;
+    trans_line->balance = double_to_gnc_numeric (0.0, xaccAccountGetCommoditySCU (account),
+                          GNC_HOW_RND_ROUND_HALF_UP);
+
+    /* We make the line_no -1 just to mark that it hasn't been set. We
+     * may get rid of line_no soon anyway, so it's not particularly
+     * important. */
+    trans_line->line_no = -1;
+
+    QofBook* book = gnc_account_get_book (account);
+    gnc_commodity* currency = xaccAccountGetCommodity (account);
+    trans_line->trans = xaccMallocTransaction (book);
+    xaccTransBeginEdit (trans_line->trans);
+    xaccTransSetCurrency (trans_line->trans, currency);
+
+    /* Go through each of the properties and edit the transaction accordingly. */
+    std::string num;
+    std::string memo;
+    std::string omemo;
+    Account *oaccount = NULL;
+    bool amount_set = false;
+    gnc_numeric amount = trans_line->balance;
+
+    for (auto prop_pair : trans_props)
+    {
+        auto type = prop_pair.first;
+        auto prop = prop_pair.second;
+        switch (type)
+        {
+        case GncTransPropType::DATE:
+            {
+                auto transdate = dynamic_cast<GncTransPropImpl<time64*>*>(prop.get())->value;
```

Comment:
That comment is mostly obsoleted by the one at 440. Following that path should get rid of the need for dynamic_casts.